File: everyday code/javascript/leetcode/54 Spiral Matrix.js
54. Spiral Matrix

Given an m x n matrix, return all elements of the matrix in spiral order.

 

Example 1:


Input: matrix = [[1,2,3],[4,5,6],[7,8,9]]
Output: [1,2,3,6,9,8,7,4,5]
Example 2:


Input: matrix = [[1,2,3,4],[5,6,7,8],[9,10,11,12]]
Output: [1,2,3,4,8,12,11,10,9,5,6,7]
 

Constraints:

m == matrix.length
n == matrix[i].length
1 <= m, n <= 10
-100 <= matrix[i][j] <= 100

var spiralOrder = function(matrix) {
    if (!matrix.length || !matrix[0].length) return [];
    let result = [];
    let top = 0, left = 0;
    let bottom = matrix.length - 1;
    let right = matrix[0].length - 1;

    while (top <= bottom && left <= right) {

        
        for (let i = left; i <= right; i++) {
            result.push(matrix[top][i]);
        }
        top++;
        for (let i = top; i <= bottom; i++) {
            result.push(matrix[i][right]);
        }
        right--;

        if(top<=bottom){
        for (let i = right; i >= left; i--) {
            result.push(matrix[bottom][i]);
        }
        bottom--
        }
        if(left<=right){
        for (let i = bottom; i >= top; i--) {
            result.push(matrix[i][left]);
        }
        left++
        }
    
    }

return result
};

File: everyday code/python/leetcode/54 Spiral Matrix.py
54. Spiral Matrix

Given an m x n matrix, return all elements of the matrix in spiral order.

 

Example 1:


Input: matrix = [[1,2,3],[4,5,6],[7,8,9]]
Output: [1,2,3,6,9,8,7,4,5]
Example 2:


Input: matrix = [[1,2,3,4],[5,6,7,8],[9,10,11,12]]
Output: [1,2,3,4,8,12,11,10,9,5,6,7]
 

Constraints:

m == matrix.length
n == matrix[i].length
1 <= m, n <= 10
-100 <= matrix[i][j] <= 100

class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        
        if not matrix or not matrix[0]:
            return []

        result = []
        top=0
        left=0
        bottom, right=len(matrix)-1,len(matrix[0])-1

        while top<=bottom and left <= right:

            for i in range(left,right+1):
                result.append(matrix[top][i])
            top+=1

            for i in range(top,bottom+1):
                result.append(matrix[i][right])
            right-=1

            if top <= bottom:
                for i in range(right,left-1,-1):
                    result.append(matrix[bottom][i])
                bottom-=1

            if left <= right:
                for i in range(bottom,top-1,-1):
                    result.append(matrix[i][left])
            left+=1

        return result    
